In [240]:
# -*- coding: utf-8 -*-

"""
Created May 7, 2022
"""

import glob
import os
import pickle

import numpy as np
import pandas as pd
from IPython.display import display_html
from sklearn.feature_selection import chi2, SelectKBest
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, OrdinalEncoder

import utils


BASED_STRUCTURED_DIR = "/dartfs/rc/nosnapshots/H/HillB-scratch/OAI/structured_data"
PREDICTIONS_CSV = "/dartfs/rc/nosnapshots/H/HillB-scratch/OAI/model_predictions/predictions_2022_12_06_08_09_19_298378.csv"
DICOM_METADATA_PATH = "./data/dicom_metadata_df.pkl"
VARIABLES_OF_INTEREST = [
    "(0008, 0070) Manufacturer",
    "(0008, 1090) Manufacturer's Model Name",
    "(0012, 0030) Clinical Trial Site ID",
    "(0018, 1000) Device Serial Number",
    "Site ID - Model",
    "hospital_site",
]
CLINICAL_WAVES_TO_FOLLOWUP = {
    "00": "Screening Visit",
    "01": "12 month Annual Visit",
    "03": "24 month Annual Visit",
    "05": "36 month Annual Visit",
    "06": "48 month Annual Visit",
    "07": "60 month Annual Visit",
    "08": "72 month Annual Visit",
    "09": "84 month Annual Visit",
    "10": "96 month Annual Visit",
    "11": "108 month Annual Visit",
}


## Structured Data

Here we combine the structured data from multiple timepoints.

In [241]:
original_dataframes: dict[str, pd.DataFrame] = {}

clinical_data_files = glob.glob(os.path.join(BASED_STRUCTURED_DIR, "AllClinical*.txt"))
xray_files = glob.glob(os.path.join(BASED_STRUCTURED_DIR, "XRay*.txt"))
enrollee_files = glob.glob(os.path.join(BASED_STRUCTURED_DIR, "Enrollees*.txt"))

combined_files = sorted(clinical_data_files + xray_files)

for filename in combined_files:
    name = os.path.basename(filename).replace(".txt", "").lower()

    df = utils.read_data_frame(os.path.join(BASED_STRUCTURED_DIR, filename))

    if df is None:
        continue

    original_dataframes[name] = df

if not len(enrollee_files):
    raise ValueError("missing Enrollee.txt from structured data")

df_site = utils.read_data_frame(enrollee_files[0])[["id", "v00site"]].astype("string")


In [242]:
dfs = {}

for side in ["left", "right"]:
    subset_cols = (
        ["id", "kooskpl", "womkpl"] if side == "left" else ["id", "kooskpr", "womkpr"]
    )

    dfs[side] = utils.concatenate_from_timepoints(
        original_dataframes,
        dataset_substring="allclinical",
        subset=subset_cols,
    )

    dfs[side] = dfs[side].merge(df_site, on="id", how="inner")
    dfs[side].rename(columns={"v00site": "hospital_site"}, inplace=True)

    utils.validate_column(
        pd.Series(dfs[side].columns),
        subset_cols + ["visit", "hospital_site"],
    )

    dfs[side].columns = [
        "id",
        "koos_pain_subscore",
        "womac_pain_subscore",
        "visit",
        "hospital_site",
    ]  # type:ignore
    dfs[side]["side"] = side

all_knee_pain_scores = pd.concat([val for val in dfs.values()])
all_knee_pain_scores["visit"] = all_knee_pain_scores["visit"].apply(
    lambda x: CLINICAL_WAVES_TO_FOLLOWUP.get(x, pd.NA)
)
all_knee_pain_scores


id koos_pain_subscore womac_pain_subscore                   visit  \
0      9000099              100.0                 0.0         Screening Visit   
1      9000099               88.9                 0.0   12 month Annual Visit   
2      9000099               80.6                 4.0   24 month Annual Visit   
3      9000099               97.2                 0.0                    <NA>   
4      9000099               83.3                 2.0   36 month Annual Visit   
...        ...                ...                 ...                     ...   
48737  9999878               94.4                 0.0   60 month Annual Visit   
48738  9999878               97.2                 0.0   72 month Annual Visit   
48739  9999878               97.2                 0.0   84 month Annual Visit   
48740  9999878               84.4                 1.0   96 month Annual Visit   
48741  9999878               88.9                 1.0  108 month Annual Visit   

      hospital_site   side  
0                 B   left  
1                 B   left  
2                 B   left  
3                 B   left  
4                 B   left  
...             ...    ...  
48737             C  right  
48738             C  right  
48739             C  right  
48740             C  right  
48741             C  right  

[97484 rows x 6 columns]

In [244]:
with open(DICOM_METADATA_PATH, "rb") as f:
    dicom_metadata = pickle.load(f)

dicom_metadata["join_id"] = dicom_metadata.index.values.astype("str")
dicom_metadata["dicom_join_id"] = dicom_metadata["(0010, 0020) Patient ID"].astype(
    "str"
)
dicom_metadata["Site ID - Model"] = dicom_metadata[
    "(0012, 0030) Clinical Trial Site ID"
].str.cat(dicom_metadata["(0008, 1090) Manufacturer's Model Name"])

dicom_metadata[
    ["join_id", "dicom_join_id"]
    + [v for v in VARIABLES_OF_INTEREST if v != "hospital_site"]
]


join_id dicom_join_id  (0008, 0070) Manufacturer  \
00700804  00700804       9659701                       AGFA   
04168501  04168501       9485404            "GE Healthcare"   
02127103  02127103       9596662  FUJI PHOTO FILM Co., ltd.   
03558001  03558001       9018489                   Swissray   
01597503  01597503       9953183                      LS100   
...            ...           ...                        ...   
03844601  03844601       9257071       FUJIFILM Corporation   
00507901  00507901       9101951            Agfa-Gevaert AG   
02256603  02256603       9297051                        NaN   
03625601  03625601       9836547       FUJIFILM Corporation   
00096003  00096003       9444196                      LS100   

         (0008, 1090) Manufacturer's Model Name  \
00700804                               ADC_51xx   
04168501                        "Definium 5000"   
02127103                                    NaN   
03558001                     ddR Formula System   
01597503                                Lumisys   
...                                         ...   
03844601                                    NaN   
00507901                               ADC_5146   
02256603                                    NaN   
03625601                                    NaN   
00096003                                Lumisys   

         (0012, 0030) Clinical Trial Site ID  \
00700804                                  58   
04168501                                  46   
02127103                                  23   
03558001                                  34   
01597503                                  10   
...                                      ...   
03844601                                  58   
00507901                                  46   
02256603                                  11   
03625601                                  10   
00096003                                  22   

         (0018, 1000) Device Serial Number       Site ID - Model  
00700804                              1134            58ADC_51xx  
04168501                               NaN     46"Definium 5000"  
02127103                               NaN                   NaN  
03558001                           S402607  34ddR Formula System  
01597503                               NaN             10Lumisys  
...                                    ...                   ...  
03844601                               NaN                   NaN  
00507901                              2205            46ADC_5146  
02256603                               NaN                   NaN  
03625601                               NaN                   NaN  
00096003                               NaN             22Lumisys  

[26520 rows x 7 columns]

## Merging Dicom Metadata

We can join the OAI structured data with the dicom metadata by joining on the patient ID.  Our variables of interest for grouping koos predictions are:

- by manufacturer
- by model
- by serial number
- by site-id
- by site-id / model

In [245]:
df_combined = dicom_metadata.merge(
    all_knee_pain_scores,
    left_on="dicom_join_id",
    right_on="id",
    how="left",
).dropna(subset=["id", "koos_pain_subscore"])

dicom_timepoint = "(0012, 0051) Clinical Trial Time Point Description"

df_combined["bin_koos"] = df_combined["koos_pain_subscore"].apply(utils.binarize_koos)
df_combined["koos_pain_subscore"].astype("string")
df_combined[dicom_timepoint] = df_combined[dicom_timepoint].astype("string")
df_combined = df_combined.dropna(subset=["bin_koos"])
df_combined = df_combined.loc[df_combined["visit"] == df_combined[dicom_timepoint]]
df_combined


(0008, 0005) Specific Character Set  \
1                               ISO_IR 100   
11                              ISO_IR 100   
28                              ISO_IR 100   
38                              ISO_IR 100   
42                                     NaN   
...                                    ...   
539975                                 NaN   
539988                                 NaN   
539998                                 NaN   
540002                                 NaN   
540012                                 NaN   

                                  (0008, 0008) Image Type  \
1                                      [DERIVED, PRIMARY]   
11                                     [DERIVED, PRIMARY]   
28                                    [ORIGINAL, PRIMARY]   
38                                    [ORIGINAL, PRIMARY]   
42      [DERIVED, PRIMARY, POST_PROCESSED, , , , , , 1...   
...                                                   ...   
539975                                                NaN   
539988                                            DERIVED   
539998                                            DERIVED   
540002                  [ORIGINAL, SECONDARY, RADIOGRAPH]   
540012                  [ORIGINAL, SECONDARY, RADIOGRAPH]   

       (0008, 0012) Instance Creation Date  \
1                                 20050513   
11                                20050513   
28                                     NaN   
38                                     NaN   
42                                     NaN   
...                                    ...   
539975                                 NaN   
539988                                 NaN   
539998                                 NaN   
540002                            20050929   
540012                            20050929   

       (0008, 0013) Instance Creation Time   (0008, 0016) SOP Class UID  \
1                                   091344    1.2.840.10008.5.1.4.1.1.1   
11                                  091344    1.2.840.10008.5.1.4.1.1.1   
28                                     NaN  1.2.840.10008.5.1.4.1.1.1.1   
38                                     NaN  1.2.840.10008.5.1.4.1.1.1.1   
42                                     NaN    1.2.840.10008.5.1.4.1.1.1   
...                                    ...                          ...   
539975                                 NaN    1.2.840.10008.5.1.4.1.1.7   
539988                                 NaN    1.2.840.10008.5.1.4.1.1.1   
539998                                 NaN    1.2.840.10008.5.1.4.1.1.1   
540002                              233900    1.2.840.10008.5.1.4.1.1.1   
540012                              233900    1.2.840.10008.5.1.4.1.1.1   

                            (0008, 0018) SOP Instance UID  \
1       1.3.6.1.4.1.21767.172.16.8.165.1158825987.343....   
11      1.3.6.1.4.1.21767.172.16.8.165.1158825987.343....   
28         1.3.6.1.4.1.21767.127.0.0.1.1377304274.0.0.1.1   
38         1.3.6.1.4.1.21767.127.0.0.1.1377304274.0.0.1.1   
42      1.3.6.1.4.1.21767.172.16.8.165.1183056852.849....   
...                                                   ...   
539975  1.3.6.1.4.1.21767.172.16.10.155.1188580343.595...   
539988     1.3.6.1.4.1.21767.127.0.0.1.1300739971.0.0.1.1   
539998     1.3.6.1.4.1.21767.127.0.0.1.1300739971.0.0.1.1   
540002  1.3.6.1.4.1.21767.172.16.8.165.1143501729.639....   
540012  1.3.6.1.4.1.21767.172.16.8.165.1143501729.639....   

       (0008, 0020) Study Date (0008, 0022) Acquisition Date  \
1                     20050513                      20050513   
11                    20050513                      20050513   
28                    20130806                      20130806   
38                    20130806                      20130806   
42                    20070611                      20070611   
...                        ...                           ...   
539975                20070822                      20070822   
539988                

We can aggregate koos scores by `Manufacterer model`.

## Koos Aggregation

Aggregation of `binarized koos` allows us to see the percentage of severe knee pain x-rays for each manufacturer.

In [247]:
koos_variables = ["koos_pain_subscore", "bin_koos"]

html_raw_tables = []
for target_variable in VARIABLES_OF_INTEREST:
    df_agg = (
        df_combined[koos_variables + [target_variable]]
        .replace("", np.nan)
        .dropna(subset=[target_variable])
        .groupby(target_variable)
        .agg(["mean", "median", "count"])
    )

    html_raw_tables.append(
        df_agg.style.set_table_attributes("style='display:inline'")
        .set_caption(target_variable)
        ._repr_html_()
    )

display_html("".join(html_raw_tables), raw=True)


(0008, 0070) Manufacturer 
 
 
   
 koos_pain_subscore 
 bin_koos 
 
 
   
 mean 
 median 
 count 
 mean 
 median 
 count 
 
 
 (0008, 0070) Manufacturer 
   
   
   
   
   
   
 
 
 
 
 "GE Healthcare" 
 88.483013 
 94.400000 
 7759 
 0.336126 
 0.000000 
 7759 
 
 
 AGFA 
 87.368271 
 94.400000 
 9455 
 0.364252 
 0.000000 
 9455 
 
 
 Agfa-Gevaert AG 
 87.975816 
 94.400000 
 7869 
 0.360147 
 0.000000 
 7869 
 
 
 FUJI PHOTO FILM Co., ltd. 
 81.182679 
 88.900000 
 2396 
 0.481219 
 0.000000 
 2396 
 
 
 FUJIFILM Corporation 
 88.121656 
 94.400000 
 7402 
 0.341124 
 0.000000 
 7402 
 
 
 GE Healthcare 
 89.328483 
 97.200000 
 811 
 0.314427 
 0.000000 
 811 
 
 
 LS100 
 85.277453 
 91.700000 
 4271 
 0.401077 
 0.000000 
 4271 
 
 
 Philips Medical Systems 
 83.130526 
 91.150000 
 190 
 0.447368 
 0.000000 
 190 
 
 
 SIEMENS 
 81.545692 
 88.900000 
 650 
 0.467692 
 0.000000 
 650 
 
 
 Swissray 
 82.619614 
 88.900000 
 8713 
 0.463790 
 0.000000 
 8713 
 
 
 

 
 (0008, 1090) Manufacturer's Model Name 
 
 
   
 koos_pain_subscore 
 bin_koos 
 
 
   
 mean 
 median 
 count 
 mean 
 median 
 count 
 
 
 (0008, 1090) Manufacturer's Model Name 
   
   
   
   
   
   
 
 
 
 
 "Definium 5000" 
 88.325206 
 94.400000 
 6927 
 0.340840 
 0.000000 
 6927 
 
 
 "Thunder Platform" 
 89.796875 
 97.200000 
 832 
 0.296875 
 0.000000 
 832 
 
 
 ADC_5146 
 87.975816 
 94.400000 
 7869 
 0.360147 
 0.000000 
 7869 
 
 
 ADC_51xx 
 87.368271 
 94.400000 
 9455 
 0.364252 
 0.000000 
 9455 
 
 
 DigitalDiagnost 
 82.902151 
 90.600000 
 186 
 0.451613 
 0.000000 
 186 
 
 
 Discovery XR656 
 89.328483 
 97.200000 
 811 
 0.314427 
 0.000000 
 811 
 
 
 FLUOROSPOT_COMPACT 
 83.300000 
 83.300000 
 2 
 0.500000 
 0.500000 
 2 
 
 
 Lumisys 
 85.432595 
 91.700000 
 3930 
 0.400763 
 0.000000 
 3930 
 
 
 SIEMENS FD-X 
 81.540278 
 88.900000 
 648 
 0.467593 
 0.000000 
 648 
 
 
 ddR Combi System 
 77.600000 
 83.300000 
 8 
 0.500000 
 0.500000 
 8 
 
 
 ddR Formula System 
 82.855338 
 88.900000 
 3269 
 0.455797 
 0.000000 
 3269 
 
 
 ddR Modulaire System 
 82.549572 
 88.900000 
 5378 
 0.467274 
 0.000000 
 5378 
 
 
 ddR Multi System 
 76.520690 
 83.300000 
 58 
 0.586207 
 1.000000 
 58 
 
 
 digital DIAGNOST 
 93.750000 
 100.000000 
 4 
 0.250000 
 0.000000 
 4 
 
 
 

 
 (0012, 0030) Clinical Trial Site ID 
 
 
   
 koos_pain_subscore 
 bin_koos 
 
 
   
 mean 
 median 
 count 
 mean 
 median 
 count 
 
 
 (0012, 0030) Clinical Trial Site ID 
   
   
   
   
   
   
 
 
 
 
 1 
 97.050000 
 97.050000 
 2 
 0.000000 
 0.000000 
 2 
 
 
 5 
 88.900000 
 88.900000 
 2 
 0.500000 
 0.500000 
 2 
 
 
 10 
 90.197833 
 97.200000 
 2492 
 0.274077 
 0.000000 
 2492 
 
 
 11 
 86.789731 
 94.400000 
 8949 
 0.370544 
 0.000000 
 8949 
 
 
 22 
 84.840659 
 91.150000 
 182 
 0.428571 
 0.000000 
 182 
 
 
 23 
 80.585016 
 87.500000 
 4418 
 0.491852 
 0.000000 
 4418 
 
 
 33 
 72.200000 
 72.200000 
 2 
 1.000000 
 1.000000 
 2 
 
 
 34 
 83.183036 
 90.600000 
 560 
 0.442857 
 0.000000 
 560 
 
 
 35 
 82.775842 
 88.900000 
 7248 
 0.462748 
 0.000000 
 7248 
 
 
 38 
 95.850000 
 95.850000 
 2 
 0.000000 
 0.000000 
 2 
 
 
 40 
 83.350000 
 89.100000 
 6 
 0.500000 
 0.500000 
 6 
 
 
 41 
 98.600000 
 98.600000 
 2 
 0.000000 
 0.000000 
 2 
 
 
 46 
 88.232712 
 94.400000 
 15618 
 0.348124 
 0.000000 
 15618 
 
 
 55 
 76.400000 
 76.400000 
 2 
 0.500000 
 0.500000 
 2 
 
 
 58 
 87.698090 
 94.400000 
 13140 
 0.356393 
 0.000000 
 13140 
 
 
 I1 
 87.150000 
 97.200000 
 8 
 0.250000 
 0.000000 
 8 
 
 
 

 
 (0018, 1000) Device Serial Number 
 
 
   
 koos_pain_subscore 
 bin_koos 
 
 
   
 mean 
 median 
 count 
 mean 
 median 
 count 
 
 
 (0018, 1000) Device Serial Number 
   
   
   
   
   
   
 
 
 
 
 08.02.366 
 100.000000 
 100.000000 
 2 
 0.000000 
 0.000000 
 2 
 
 
 08.02.399 
 87.500000 
 87.500000 
 2 
 0.500000 
 0.500000 
 2 
 
 
 1003 
 83.300000 
 83.300000 
 2 
 0.500000 
 0.500000 
 2 
 
 
 

## Model predictions

These predictions are generated by the `koos` deep learning model.  They are written to the path indicated by `PREDICTIONS_CSV`.  Xrays were scored using a corresponding slurm model creating a `koos` numeric prediction and a binned prediction mapped to either `0` or `1`.

In [248]:
df_predictions = pd.read_csv(
    PREDICTIONS_CSV,
    names=["batch", "koos_prediction", "patient_id", "bin_koos_prediction"],
    header=None,
    skiprows=1,
)

df_predictions["join_id"] = df_predictions["patient_id"].map("{:08}".format)
df_predictions


batch  koos_prediction  patient_id  bin_koos_prediction   join_id
0          0        97.930750     1652703                  0.0  01652703
1          1        80.273544     2002706                  1.0  02002706
2          2        91.560130     1623104                  0.0  01623104
3          3        83.731750     4181102                  1.0  04181102
4          4        86.088936     2300601                  1.0  02300601
...      ...              ...         ...                  ...       ...
26515      3        82.593605     2063801                  1.0  02063801
26516      4        98.483200     2338901                  0.0  02338901
26517      5        94.040940     1371103                  0.0  01371103
26518      6        82.667530     1632904                  1.0  01632904
26519      7        82.701420     1703104                  1.0  01703104

[26520 rows x 5 columns]

## Merging Predictions

Model predictions are merged with dicom metadata by joining again on patient ID.  This allows us to have both the actual and predicted koos value for a given xray.  The sides for each xray are included in the analysis.

In [249]:
koos_variables = ["koos_pain_subscore", "bin_koos"]

df_combined_predictions = df_combined.merge(
    df_predictions, left_on="join_id", right_on="join_id", how="inner"
).dropna(subset=["koos_prediction"] + VARIABLES_OF_INTEREST)

df_combined_predictions[
    ["join_id", "visit", "side"] + koos_variables + VARIABLES_OF_INTEREST
]


join_id                  visit   side koos_pain_subscore  bin_koos  \
0      00700804  12 month Annual Visit   left               80.6       1.0   
1      00700804  12 month Annual Visit  right              100.0       0.0   
6      03558001  72 month Annual Visit   left              100.0       0.0   
7      03558001  72 month Annual Visit  right              100.0       0.0   
16     02102701  36 month Annual Visit   left               88.9       0.0   
...         ...                    ...    ...                ...       ...   
52962  02143601  36 month Annual Visit  right               97.2       0.0   
52971  00333904        Screening Visit   left              100.0       0.0   
52972  00333904        Screening Visit  right               61.1       1.0   
52975  00507901        Screening Visit   left              100.0       0.0   
52976  00507901        Screening Visit  right              100.0       0.0   

      (0008, 0070) Manufacturer (0008, 1090) Manufacturer's Model Name  \
0                          AGFA                               ADC_51xx   
1                          AGFA                               ADC_51xx   
6                      Swissray                     ddR Formula System   
7                      Swissray                     ddR Formula System   
16                         AGFA                               ADC_51xx   
...                         ...                                    ...   
52962                      AGFA                               ADC_51xx   
52971                      AGFA                               ADC_51xx   
52972                      AGFA                               ADC_51xx   
52975           Agfa-Gevaert AG                               ADC_5146   
52976           Agfa-Gevaert AG                               ADC_5146   

      (0012, 0030) Clinical Trial Site ID (0018, 1000) Device Serial Number  \
0                                      58                              1134   
1                                      58                              1134   
6                                      34                           S402607   
7                                      34                           S402607   
16                                     58                              1522   
...                                   ...                               ...   
52962                                  58                              1134   
52971                                  58                              1134   
52972                                  58                              1134   
52975                                  46                              2205   
52976                                  46                              2205   

            Site ID - Model hospital_site  
0                58ADC_51xx             D  
1                58ADC_51xx             D  
6      34ddR Formula System             A  
7      34ddR Formula System             A  
16               58ADC_51xx             D  
...                     ...           ...  
52962            58ADC_51xx             D  
52971            58ADC_51xx             D  
52972            58ADC_51xx             D  
52975            46ADC_5146             C  
52976            46ADC_5146             C  

[26483 rows x 11 columns]

In [250]:
def prediction_performance(ytrue, ypred):
    stats = {}

    try:
        cm = confusion_matrix(ytrue, ypred)

        spec_sum = cm[1, 0] + cm[1, 1]
        sen_sum = cm[0, 0] + cm[0, 1]
        prec_sum = cm[0, 0] + cm[1, 0]

        if spec_sum == 0 or sen_sum == 0 or prec_sum == 0:
            raise Exception

        if not len(ypred) or len(cm) <= 1:
            raise Exception

        stats["Accuracy"] = (cm[0, 0] + cm[1, 1]) / cm.sum()
        stats["Precision"] = cm[0, 0] / (cm[0, 0] + cm[1, 0])
        stats["Specificity"] = cm[1, 1] / (spec_sum)
        stats["Sensitivity"] = cm[0, 0] / (sen_sum)
        stats["ER"] = (cm[0, 1] + cm[1, 0]) / cm.sum()
        stats["FPR"] = 1 - stats["Specificity"]
        stats["TP Count"] = cm[0, 0]
        stats["TN Count"] = cm[1, 1]
    except:
        stats["Accuracy"] = np.nan
        stats["Precision"] = np.nan
        stats["Specificity"] = np.nan
        stats["Sensitivity"] = np.nan
        stats["ER"] = np.nan
        stats["FPR"] = np.nan
        stats["TP Count"] = np.nan
        stats["TN Count"] = np.nan

    return stats


global_ytrue = df_combined_predictions["bin_koos"]
global_ypred = df_combined_predictions["bin_koos_prediction"]
global_cm = confusion_matrix(global_ytrue, global_ypred)

prediction_performance(global_ytrue, global_ypred)


{'Accuracy': 0.5068156930861307,
 'Precision': 0.5996302382908792,
 'Specificity': 0.4437785388127854,
 'Sensitivity': 0.5483063051781354,
 'ER': 0.4931843069138693,
 'FPR': 0.5562214611872146,
 'TP Count': 8757,
 'TN Count': 4665}

## Global Performance

The global performance of prediction gives an accuracy of 0.507 with a precision of 0.600.  Sensitivity is 0.548 and specificity is 0.444.

## Koos Accuracy by Manufacturer

Grouping the predictions by manufacturer shows varying levels of accuracy.  The most accurately predicted pain scores occured when the manufacturer model was the `SIEMENS`.  The least accurate occurred with `Agfa-Gevaert AG`.

## Accuracy Discrepancies

Many of the xray manufacturers show widely varying sensitivities and specificities.  Aside from the large drop in accuracy with `Agfa-Gevaert AG`, there is a large difference between sensitivity even at similar counts.  These differences also hold for model names and site id.

In [251]:
def grouping_func(row):
    ytrue = row["bin_koos"]
    ypred = row["bin_koos_prediction"]
    d = {"Count": len(ypred)}
    d.update(prediction_performance(ytrue, ypred))

    return pd.Series(d)


html_raw_tables = []
for grouping_var in VARIABLES_OF_INTEREST:
    df_combined_predictions[grouping_var] = df_combined_predictions[
        grouping_var
    ].astype("string")
    df_agg = (
        df_combined_predictions[df_combined_predictions[grouping_var] != ""]
        .groupby([grouping_var])
        .apply(grouping_func)
        .sort_values(by=["Count"], ascending=False)
    )

    html_raw_tables.append(
        df_agg.style.set_table_attributes("style='display:inline'")
        .set_caption(grouping_var)
        ._repr_html_()
    )

display_html("".join(html_raw_tables), raw=True)


(0008, 0070) Manufacturer 
 
 
   
 Count 
 Accuracy 
 Precision 
 Specificity 
 Sensitivity 
 ER 
 FPR 
 TP Count 
 TN Count 
 
 
 (0008, 0070) Manufacturer 
   
   
   
   
   
   
   
   
   
 
 
 
 
 AGFA 
 9343.000000 
 0.551857 
 0.647763 
 0.373187 
 0.653085 
 0.448143 
 0.626813 
 3895.000000 
 1261.000000 
 
 
 Swissray 
 8421.000000 
 0.525828 
 0.542532 
 0.290644 
 0.729874 
 0.474172 
 0.709356 
 3291.000000 
 1137.000000 
 
 
 Agfa-Gevaert AG 
 7867.000000 
 0.427482 
 0.637922 
 0.753884 
 0.243893 
 0.572518 
 0.246116 
 1228.000000 
 2135.000000 
 
 
 SIEMENS 
 650.000000 
 0.563077 
 0.564583 
 0.312500 
 0.783237 
 0.436923 
 0.687500 
 271.000000 
 95.000000 
 
 
 Philips Medical Systems 
 190.000000 
 0.552632 
 0.586207 
 0.435294 
 0.647619 
 0.447368 
 0.564706 
 68.000000 
 37.000000 
 
 
 GE Healthcare 
 12.000000 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 
 
 

 
 (0008, 1090) Manufacturer's Model Name 
 
 
   
 Count 
 Accuracy 
 Precision 
 Specificity 
 Sensitivity 
 ER 
 FPR 
 TP Count 
 TN Count 
 
 
 (0008, 1090) Manufacturer's Model Name 
   
   
   
   
   
   
   
   
   
 
 
 
 
 ADC_51xx 
 9343.000000 
 0.551857 
 0.647763 
 0.373187 
 0.653085 
 0.448143 
 0.626813 
 3895.000000 
 1261.000000 
 
 
 ADC_5146 
 7867.000000 
 0.427482 
 0.637922 
 0.753884 
 0.243893 
 0.572518 
 0.246116 
 1228.000000 
 2135.000000 
 
 
 ddR Modulaire System 
 5100.000000 
 0.523725 
 0.536206 
 0.252926 
 0.762925 
 0.476275 
 0.747074 
 2066.000000 
 605.000000 
 
 
 ddR Formula System 
 3261.000000 
 0.529592 
 0.555658 
 0.352624 
 0.677746 
 0.470408 
 0.647376 
 1203.000000 
 524.000000 
 
 
 SIEMENS FD-X 
 648.000000 
 0.563272 
 0.564854 
 0.313531 
 0.782609 
 0.436728 
 0.686469 
 270.000000 
 95.000000 
 
 
 DigitalDiagnost 
 186.000000 
 0.548387 
 0.580357 
 0.440476 
 0.637255 
 0.451613 
 0.559524 
 65.000000 
 37.000000 
 
 
 ddR Multi System 
 52.000000 
 0.538462 
 0.476190 
 0.266667 
 0.909091 
 0.461538 
 0.733333 
 20.000000 
 8.000000 
 
 
 Discovery XR656 
 12.000000 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 
 
 ddR Combi System 
 8.000000 
 0.250000 
 0.333333 
 0.000000 
 0.500000 
 0.750000 
 1.000000 
 2.000000 
 0.000000 
 
 
 digital DIAGNOST 
 4.000000 
 0.750000 
 0.750000 
 0.000000 
 1.000000 
 0.250000 
 1.000000 
 3.000000 
 0.000000 
 
 
 FLUOROSPOT_COMPACT 
 2.000000 
 0.500000 
 0.500000 
 0.000000 
 1.000000 
 0.500000 
 1.000000 
 1.000000 
 0.000000 
 
 
 

 
 (0012, 0030) Clinical Trial Site ID 
 
 
   
 Count 
 Accuracy 
 Precision 
 Specificity 
 Sensitivity 
 ER 
 FPR 
 TP Count 
 TN Count 
 
 
 (0012, 0030) Clinical Trial Site ID 
   
   
   
   
   
   
   
   
   
 
 
 
 
 58 
 9351.000000 
 0.551706 
 0.647997 
 0.373002 
 0.652771 
 0.448294 
 0.626998 
 3899.000000 
 1260.000000 
 
 
 46 
 7859.000000 
 0.427535 
 0.637922 
 0.753623 
 0.244135 
 0.572465 
 0.246377 
 1228.000000 
 2132.000000 
 
 
 35 
 7194.000000 
 0.528774 
 0.545489 
 0.287173 
 0.736869 
 0.471226 
 0.712827 
 2848.000000 
 956.000000 
 
 
 23 
 1467.000000 
 0.534424 
 0.541459 
 0.345221 
 0.707572 
 0.465576 
 0.654779 
 542.000000 
 242.000000 
 
 
 34 
 556.000000 
 0.535971 
 0.561275 
 0.278226 
 0.743506 
 0.464029 
 0.721774 
 229.000000 
 69.000000 
 
 
 22 
 40.000000 
 0.250000 
 0.291667 
 0.150000 
 0.350000 
 0.750000 
 0.850000 
 7.000000 
 3.000000 
 
 
 40 
 6.000000 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 
 
 5 
 2.000000 
 0.500000 
 0.500000 
 0.000000 
 1.000000 
 0.500000 
 1.000000 
 1.000000 
 0.000000 
 
 
 33 
 2.000000 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 
 
 38 
 2.000000 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 
 
 41 
 2.000000 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 nan 
 
 
 55 
 2.000000 
 0.500000 
 0.500000 
 0.000000 
 1.000000 
 0.500000 
 1.000000 
 1.000000 
 0.000000 
 
 
 

 
 (0018, 1000) Device Serial Number 
 
 
   
 Count 
 Accuracy 
 Precision 
 Specificity 
 Sensitivity 
 ER 
 FPR 
 TP

## Feature Selection Analysis

In [252]:
from sklearn.preprocessing import LabelEncoder, OrdinalEncoder
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import chi2, SelectKBest


TEST_VARIABLES = [
    "(0008, 0080) Institution Name",
    "(0008, 0008) Image Type",
    "(0008, 1030) Study Description",
    "(0018, 1000) Device Serial Number",
    "(0018, 1020) Software Versions",
    "(0018, 5101) View Position",
    *VARIABLES_OF_INTEREST,
]


def feature_select(dataset, yvar="bin_koos", xvars=[]):
    if not len(xvars):
        raise ValueError("xvars cannot be empty")

    le = LabelEncoder()
    oe = OrdinalEncoder()

    X = dataset[xvars].astype(str)
    y = dataset[yvar].values

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.33, random_state=0
    )

    le.fit(y_train)
    y_train_enc = le.transform(y_train)
    y_test_enc = le.transform(y_test)

    oe.fit(X_train)
    X_train_enc = oe.transform(X_train)
    X_test_enc = oe.transform(X_test)

    fs = SelectKBest(score_func=chi2, k="all")
    fs.fit(X_train_enc, y_train_enc)

    cols = fs.get_support(indices=True)
    feature_names = X.iloc[:, cols].columns

    X_train_fs = fs.transform(X_train_enc)
    X_test_fs = fs.transform(X_test_enc)

    for i in range(len(fs.scores_)):
        print("Feature %s: %f" % (feature_names[i], fs.scores_[i]))


feature_select(df_combined_predictions, yvar="bin_koos", xvars=TEST_VARIABLES)


Feature (0008, 0080) Institution Name: 20.746257
Feature (0008, 0008) Image Type: 525.462432
Feature (0008, 1030) Study Description: 67.581534
Feature (0018, 1000) Device Serial Number: 34.454259
Feature (0018, 1020) Software Versions: 17.500046
Feature (0018, 5101) View Position: 0.422160
Feature (0008, 0070) Manufacturer: 391.664286
Feature (0008, 1090) Manufacturer's Model Name: 643.504002
Feature (0012, 0030) Clinical Trial Site ID: 161.297454
Feature (0018, 1000) Device Serial Number: 34.454259
Feature Site ID - Model: 197.672877
Feature hospital_site: 30.200640


In [253]:
pd.set_option("display.max_rows", 75)

target_variable = "hospital_site"
model_name = "(0008, 1090) Manufacturer's Model Name"

df_combined.groupby([target_variable, model_name])["id"].agg(["count"])


count
hospital_site (0008, 1090) Manufacturer's Model Name       
A                                                         0
              "Definium 5000"                             4
              "Thunder Platform"                          2
              ADC_5146                                    4
              ADC_51xx                                    8
              DigitalDiagnost                           128
              Discovery XR656                             0
              FLUOROSPOT_COMPACT                          0
              Lumisys                                     2
              SIEMENS FD-X                                0
              ddR Combi System                            8
              ddR Formula System                       2466
              ddR Modulaire System                     5378
              ddR Multi System                           58
              digital DIAGNOST                            2
B                                                       203
              "Definium 5000"                             0
              "Thunder Platform"                          0
              ADC_5146                                    0
              ADC_51xx                                    0
              DigitalDiagnost                             0
              Discovery XR656                             0
              FLUOROSPOT_COMPACT                          0
              Lumisys                                  3568
              SIEMENS FD-X                                0
              ddR Combi System                            0
              ddR Formula System                          0
              ddR Modulaire System                        0
              ddR Multi System                            0
              digital DIAGNOST                            0
C                                                         0
              "Definium 5000"                          6923
              "Thunder Platform"                        830
              ADC_5146                                 7865
              ADC_51xx                                    0
              DigitalDiagnost                             0
              Discovery XR656                             0
              FLUOROSPOT_COMPACT                          0
              Lumisys                                     0
              SIEMENS FD-X                                0
              ddR Combi System                            0
              ddR Formula System                          0
              ddR Modulaire System                        0
              ddR Multi System                            0
              digital DIAGNOST                            0
D                                                         0
              "Definium 5000"                             0
              "Thunder Platform"                          0
              ADC_5146                                    0
              ADC_51xx                                 9447
              DigitalDiagnost                             0
              Discovery XR656                           811
              FLUOROSPOT_COMPACT                          0
              Lumisys                                     0
              SIEMENS FD-X                                0
              ddR Combi System                            0
              ddR Formula System                          0
              ddR Modulaire System                        0
              ddR Multi System                            0
              digital DIAGNOST                            0
E                                                      1600
              "Definium 5000"                             0
              "Thunder Platform"                          0
              ADC_5146                                    0
              ADC_51xx                                    0
              DigitalDiagnost     